### DISCUSSION TITLE
another real-world-example (with --interactive option)

### DISCUSSION THREAD

**Discussion by felixlohmeier** [bashly-framework/bashly#390] [Show and tell]
Hi Danny,

We had a discussion on catch-all-advanced #199 some time ago and I would now like to report back that your advice to use repeatable args worked well. I used Bashly to create a command line client for OpenRefine: https://github.com/opencultureconsulting/orcli/blob/main/src/bashly.yml. The client has many commands and options and Bashly has helped a lot in keeping the code structured and maintainable.

Of general interest to the Bashly community might be the `--interactive` option. I'm not sure if my solution works well in all scenarios, but for now it seems to work fine. With `./orcli run --interactive` the user gets a shell with autocompletion and alias without prior installation. Maybe this is an idea for another Bashly library?

https://github.com/opencultureconsulting/orcli/blob/6c47e108f7b978a030e57bd05e7421fece5813ab/src/run_command.sh#L84-L96
```
if [[ ${args[--interactive]} ]]; then
    # case 2: execute scripts and keep shell running
    bash --rcfile <(
        cat ~/.bashrc
        if ! command -v orcli &>/dev/null; then
            echo "alias orcli=${scriptpath}/orcli"
        fi
        for i in "${!files[@]}"; do
            log "executing script ${files[$i]}..."
            awk 1 "${files[$i]}"
        done
        interactive
    ) -i </dev/tty
(...)
```

https://github.com/opencultureconsulting/orcli/blob/56272613f2791e76aa890516ea84ca2a5c3adc47/src/lib/interactive.sh
```
function interactive() {
    cat <<'EOF'
PS1="(orcli) [\u@\h \W]\$ "
source <(orcli completions)
echo '================================================================'
echo 'Interactive Bash shell with OpenRefine running in the background'
echo 'Use the "orcli" command and tab completion to control OpenRefine'
echo 'Type "history -a FILE" to write out your session history'
echo 'Type "exit" or CTRL-D to destroy temporary OpenRefine workspace'
echo '================================================================'
EOF
}
```

Thanks again for your advice!

**Comment by DannyBen** [comment]
Nice. Thanks for sharing.

1. Can I add your CLI to the [real world examples](https://bashly.dannyb.co/examples/#real-world-examples)?
2. I will take a deeper look at your interactive implementation. I like interactive terminals. In fact, bashly uses my mister_bin gem, which also [comes with an interactive terminal](https://github.com/DannyBen/mister_bin#interactive-terminal) feature.

**Comment by DannyBen** [comment]
Alright. So here are my thoughts:

### Generic interactive terminal

When I think about an interactive terminal, I think of something like the redis CLI or the Postgres CLI: In both, you enter their own "world" with their own commands. It is **not** a bash subshell.

To create such an experience for any command, we can use `rlwrap` as the below bash script demonstrates:

```bash
#!/usr/bin/env bash
if [[ $# -eq 0 ]]; then
  echo "Interactive shell"
  echo "  Usage:   $0 COMMAND"
  echo "  Example: $0 git"
  exit 1
fi

command="$1"

rlwrap \
  --always-readline \
  --complete-filenames \
  --no-children \
  --prompt-colour=Yellow \
  --substitute-prompt "$command > " \
  -t dumb \
  bash -c "while read -e line; do "$command" \$line || true; done"
```

running it with any command, will start a shell, where the specified command is pre-populated always.

I called this script `it` (for interactive terminal), here is how it looks:
![cast](https://user-images.githubusercontent.com/2405099/236669558-6fac89c5-297a-4ed6-bb31-718b69125d52.gif)

I am posting it here, thinking perhaps it would be helpful for your users, using your shell.

### Built in `bashly add shell`

In regards to adding something like this (or like the approach you took) as a standard library - I wonder if it is a common enough use case, and if it can be done in a generic way. Your approach - as I understand it - adds things specific to your use case, and furthermore, requires to have completions in order to provide some value.

I wonder if this could (or should) be generalized.

### Adding a `bashly shell` command

As I mentioned, the command line library that bashly itself uses, supports creating an interactive shell of the command it describes.

So I went ahead and created a `bashly shell` command (#391), which I think could be useful. In it, you can run any bashly command with these advantages:

1. You do not need to type `bashly` repeatidly
2. It is faster, since it does not re-invoke Ruby

Here is a quick example of how it works.
![cast](https://user-images.githubusercontent.com/2405099/236669374-b08bf683-715b-4585-a681-30fc10d90014.gif)


Thoughts?

**Comment by felixlohmeier** [comment]
> Can I add your CLI to the [real world examples](https://bashly.dannyb.co/examples/#real-world-examples)?

Sure, thanks a lot.
 
> Generic interactive terminal
> When I think about an interactive terminal, I think of something like the redis CLI or the Postgres CLI: In both, you enter their own "world" with their own commands. It is **not** a bash subshell.

True, this is probably the much more common use case and user expectation.

To illustrate the orcli use case: A full bash shell should indeed be available so that other unix tools can be used. The flow would be something like this:
1. user enters "orcli run --interactive": orcli starts OpenRefine as a background service with a temporary directory and provides a shell with orcli autocompletion.
2. user plays with orcli and also uses bash for loop, &/wait for parallelization and other unix tools like wc, time etc.
3. user terminates shell: orcli takes care of killing OpenRefine and deleting the temporary directories.

**Comment by DannyBen** [comment]
- [Added orcli to examples](https://bashly.dannyb.co/examples/#real-world-examples).
- I will merge the [bashly shell PR](https://github.com/DannyBen/bashly/pull/391).
